docs/enhancements/agent-late-binding.md

```diff
@@ -0,0 +1,125 @@
+
+- For SaaS, there are discussions about changing the tenancy model to include organizations and not only users.  Those decisions are out of scope for this document.
+
+## Proposal
+
+There are two main (non-breaking) changes to the CRDs:
+1. ClusterRef in the InfraEnv Spec will no longer be mandatory.
+1. ClusterRef in the Agent Spec will no longer be mandatory.
+
+A Discovery ISO with no associated cluster means that hosts booted with that ISO will have agents that are not bound to any cluster.  Thus, Agent CRs can be created with no cluster binding.  It will still be possible to retain the previous behavior by providing a ClusterRef.  The multitenancy model will remain unchanged, with Agents being created in the namespace of the InfraEnv they booted from.
+
+The Agent's ClusterRef may now be deleted or changed dynamically, unless the Agent is part of an ongoing cluster installation, in which case the installation must first be cancelled (not currently possible via the k8s API).  If the host had started/completed its installation, a new Agent Condition would indicate that the user needs to boot the host into the Discovery ISO once
+again (as mentioned in the non-goals, this can be done automatically but is not in the scope of this enhancement).
+
+The REST API will require bigger changes:
+1. Move images to be a standalone resource, rather than part of the cluster resource (currently /clusters/{cluster_id}/downloads/image).  This removes the dependency of an image on a specific cluster, removes the current limitation of one image per cluster, and aligns with InfraEnv CRD.  The image resource will have an additional optional cluster-id property.
+1. Move hosts to be a standalone resource, rather than under of the cluster resource (currently /clusters/{cluster_id}/hosts/{host_id}), and add add an optional cluster-id property.
+
+For the SaaS, the service will record ownership (user and org ID) for images, and hosts will inherit the same ownership.
+
+Agent validations that do not depend on being part of a cluster will run for Agents that are both associated with clusters and those that are not.  Examples of these include minimum hardware for any role, synced NTP, and registry access.  Validations that do depend on being part of a cluster, such as connectivity checks, will run only for Agents associated with clusters.
+
+Events are currently always scoped to a cluster.  We will now need two types of events:
+1. The events that we have today, which should be visible to the Cluster Creator.
+1. The events for an Agent which is not associated with a cluster, for example a history of when it registered with the pool and which clusters it was associated with (should be visible to the infrastructure admin).  We will also want to be able to see the events for all Agents which we have permissions for.
+
```

Comment:
What about Authorization? Currently local auth uses the clusterID

Reply:
Agreed, we need to think about this.
Presumably the first time the service is used we will allow agents to register without any clusters in the system yet.

Some background though...
Local auth doesn't use the cluster id for authz, it's for token revocation (authn). So really we may just need to create a "real" token tracking system along with this.

Reply:
I assume we considered adding a TTL to the token?

Reply:
The token is baked into the iso and can't be renewed so that wouldn't work for the case where an iso is pre-installed on some hardware or if the machine is just sitting in a pool for a while.

Reply:
I think that the "pool-id" can just replace the cluster-id for the token

Reply:
Hopefully we won't have to revoke the token. Would happen if we had to delete the pool "cluster"?

Reply:
I wouldn't want to force someone to delete a "pool" to revoke a token.  A cluster is short-lived, that's fine.  A pool isn't.

Reply:
So what are we suggesting to use for the token if not the pool-id?

Reply:
I don't have a good answer at the moment

Reply:
maybe we need an entity for 'Token' that will have its own ID and we will create tokens and will be able to delete/revoke them.

Reply:
I think that's what @carbonin meant by "we may just need to create a "real" token tracking system along with this".  We can, then we'll probably create an auth service.  Better to use an existing service like keystone, vault, or similar.

Reply:
I'm still shaken that you mentioned keystone